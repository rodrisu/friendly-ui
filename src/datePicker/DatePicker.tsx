import React, { PureComponent } from 'react'
import DayPicker, { CaptionElementProps, DayModifiers, NavbarElementProps } from 'react-day-picker'
import cc from 'classcat'

import { prefix } from '../_utils'
import { color } from '../_utils/branding'
import { OnChangeParameters } from '../_utils/onChange'
import { Button, ButtonStatus } from '../button'
import { ArrowIcon } from '../icon/arrowIcon'
import { StyledDatePicker } from './DatePicker.style'

const BASE_CLASSNAME = 'datepicker'

const defaultWeekdaysLong = [0, 1, 2, 3, 4, 5, 6].map(weekday =>
  DayPicker.LocaleUtils.formatWeekdayLong(weekday),
)
const defaultWeekdaysShort = [0, 1, 2, 3, 4, 5, 6].map(weekday =>
  DayPicker.LocaleUtils.formatWeekdayShort(weekday),
)
const defaultMonths = DayPicker.LocaleUtils.getMonths()

export enum DatePickerOrientation {
  HORIZONTAL = 'horizontal',
  VERTICAL = 'vertical',
}

export type DatePickerProps = Readonly<{
  name: string
  locale?: string
  weekdaysShort?: Array<string>
  weekdaysLong?: Array<string>
  months?: Array<string>
  onChange?: (obj: OnChangeParameters) => void
  initialDate?: Date
  initialMonth?: Date
  className?: string
  numberOfMonths?: number
  orientation?: DatePickerOrientation
  isOutsideRange?: (day: Date) => boolean
  fromMonth?: Date
  toMonth?: Date
  firstDayOfWeek?: number
  stickyPositionTop?: number
  focus?: boolean
}>

export type DatePickerState = {
  date: Date
}

export class DatePicker extends PureComponent<DatePickerProps, DatePickerState> {
  static defaultProps: Partial<DatePickerProps> = {
    locale: 'en',
    weekdaysShort: defaultWeekdaysShort,
    weekdaysLong: defaultWeekdaysLong,
    months: defaultMonths,
    numberOfMonths: 2,
    initialDate: null,
    initialMonth: new Date(),
    isOutsideRange: day => DayPicker.DateUtils.isDayBefore(day, new Date()),
    orientation: DatePickerOrientation.HORIZONTAL,
    fromMonth: new Date(),
    toMonth: new Date(new Date().getFullYear() + 1, new Date().getMonth(), new Date().getDate()),
    firstDayOfWeek: 0,
    stickyPositionTop: 0,
    focus: false,
  }

  state = {
    date: this.props.initialDate,
  }

  dayPickerContainer = React.createRef<HTMLDivElement>()

  formatMonthTitle = (date: Date): string => {
    const currentYear = `${new Date().getFullYear()}`
    return `${this.props.months[date.getMonth()]} ${date.getFullYear()}`.replace(currentYear, '')
  }

  onDayClick = (date: Date, modifiers: DayModifiers) => {
    if (!modifiers.disabled) {
      this.setState({ date })

      if (this.props.onChange) {
        const yearString = `${date.getFullYear()}`
        const monthString = `${date.getMonth() + 1}`.padStart(2, '0')
        const dayString = `${date.getDate()}`.padStart(2, '0')
        this.props.onChange({
          name: this.props.name,
          value: `${yearString}-${monthString}-${dayString}`,
        })
      }
    }
  }

  renderNavbar = (props: NavbarElementProps) => {
    if (this.props.orientation === DatePickerOrientation.VERTICAL) {
      // re-order weekdays starting from given firstDayOfWeek
      const orderedWeekdays = [
        ...this.props.weekdaysShort.slice(this.props.firstDayOfWeek),
        ...this.props.weekdaysShort.slice(0, this.props.firstDayOfWeek),
      ]
      const style = { top: `${this.props.stickyPositionTop}px` }
      return (
        <div
          className={cc(prefix({ 'sticky-weekdays': true }, BASE_CLASSNAME))}
          style={style}
          aria-hidden="true"
        >
          <div className={cc(prefix({ 'sticky-weekdaysrow': true }, BASE_CLASSNAME))}>
            {orderedWeekdays.map(weekday => (
              <div className={cc(prefix({ 'sticky-weekday': true }, BASE_CLASSNAME))} key={weekday}>
                {weekday}
              </div>
            ))}
          </div>
        </div>
      )
    }

    return (
      <div className={props.className}>
        {props.showPreviousButton && (
          <Button
            status={ButtonStatus.TERTIARY}
            className={prefix({ 'previous-month': true }, BASE_CLASSNAME)}
            onClick={() => props.onPreviousClick()}
            isBubble
          >
            <ArrowIcon iconColor={color.blue} />
          </Button>
        )}
        {props.showNextButton && (
          <Button
            status={ButtonStatus.TERTIARY}
            className={prefix({ 'next-month': true }, BASE_CLASSNAME)}
            onClick={() => props.onNextClick()}
            isBubble
          >
            <ArrowIcon right iconColor={color.blue} />
          </Button>
        )}
      </div>
    )
  }

  renderCaption = (props: CaptionElementProps) => (
    <div className={cc([prefix({ 'month-caption': true }, BASE_CLASSNAME), 'DayPicker-Caption'])}>
      {props.localeUtils.formatMonthTitle(props.date)}
    </div>
  )

  renderDay = (day: Date) => <span>{day.getDate()}</span>

  scrollToSelectedMonth = () => {
    const { initialDate, initialMonth } = this.props

    if (initialDate === null) {
      return
    }

    // Get the number of months between initialMonth (first month) and initialDate (selected date)
    // This gives us the index to select the right month in '.DayPicker-Month' array
    const month =
      (initialDate.getFullYear() - initialMonth.getFullYear()) * 12 +
      initialDate.getMonth() -
      initialMonth.getMonth()

    const selectedMonth = this.dayPickerContainer.current.querySelectorAll('.DayPicker-Month')[
      month
    ] as HTMLElement

    if (typeof selectedMonth === 'undefined') {
      return
    }

    const top = selectedMonth.offsetTop + this.props.stickyPositionTop
    if (top > window.innerHeight + window.pageYOffset) {
      window.scrollTo(0, top)
    }
  }

  componentDidMount() {
    this.scrollToSelectedMonth()
    if (this.dayPickerContainer.current && this.props.focus) {
      const datePickerFocusableElement = this.dayPickerContainer.current.querySelector(
        '.DayPicker-wrapper',
      ) as HTMLElement
      if (datePickerFocusableElement) {
        datePickerFocusableElement.focus()
      }
    }
  }

  componentDidUpdate() {
    this.scrollToSelectedMonth()
  }

  render() {
    const {
      className,
      numberOfMonths,
      isOutsideRange,
      orientation,
      fromMonth,
      toMonth,
      weekdaysShort,
      weekdaysLong,
      months,
      firstDayOfWeek,
      initialMonth,
    } = this.props
    const { date } = this.state
    const layoutClassName = `months-grid-${numberOfMonths}`

    return (
      <StyledDatePicker
        ref={this.dayPickerContainer}
        className={cc([
          `kirk-${BASE_CLASSNAME}`,
          className,
          prefix({ [orientation]: true }, BASE_CLASSNAME),
          prefix({ [layoutClassName]: true }, BASE_CLASSNAME),
        ])}
      >
        <DayPicker
          numberOfMonths={numberOfMonths}
          onDayClick={this.onDayClick}
          selectedDays={date}
          disabledDays={isOutsideRange}
          fromMonth={fromMonth}
          pagedNavigation
          navbarElement={this.renderNavbar}
          captionElement={this.renderCaption}
          renderDay={this.renderDay}
          weekdaysShort={weekdaysShort}
          weekdaysLong={weekdaysLong}
          months={months}
          firstDayOfWeek={firstDayOfWeek}
          localeUtils={{ ...DayPicker.LocaleUtils, formatMonthTitle: this.formatMonthTitle }}
          initialMonth={initialMonth}
          toMonth={toMonth}
        />
      </StyledDatePicker>
    )
  }
}
