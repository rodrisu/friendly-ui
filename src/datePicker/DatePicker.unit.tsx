import React from 'react'
import DayPicker, { CaptionElementProps, NavbarElementProps } from 'react-day-picker'
import renderer from 'react-test-renderer'
import { mount, shallow } from 'enzyme'

import { DatePicker, DatePickerOrientation } from './DatePicker'

describe('DatePicker', () => {
  it('Should apply the base className ', () => {
    const datepicker = shallow(<DatePicker />)
    expect(datepicker.prop('className')).toContain('kirk-datepicker')
  })

  describe('renderNavbar', () => {
    const navbarProps: Partial<NavbarElementProps> = {
      className: '',
      showNextButton: true,
      showPreviousButton: true,
      onPreviousClick() {},
      onNextClick() {},
    }
    it('Should render the weekdays in vertical mode', () => {
      const datepicker = shallow(
        <DatePicker name="datepicker" orientation={DatePickerOrientation.VERTICAL} />,
      )
      const navbar = renderer.create(datepicker.instance().renderNavbar(navbarProps))
      expect(navbar).toMatchSnapshot()
    })

    it('Should render the previous/next buttons in horizontal mode', () => {
      const datepicker = shallow(
        <DatePicker name="datepicker" orientation={DatePickerOrientation.HORIZONTAL} />,
      )
      const navbar = renderer.create(datepicker.instance().renderNavbar(navbarProps))
      expect(navbar).toMatchSnapshot()
    })
  })

  describe('renderCaption', () => {
    const currentYear = new Date().getFullYear()
    const datepicker = shallow(<DatePicker name="datepicker" />)
    const captionProps: Partial<CaptionElementProps> = {
      date: new Date(currentYear, 0, 1),
      localeUtils: {
        ...DayPicker.LocaleUtils,
        formatMonthTitle: datepicker.instance().formatMonthTitle,
      },
    }
    it('Should render the given month title', () => {
      const caption = renderer.create(datepicker.instance().renderCaption(captionProps))
      expect(caption).toMatchSnapshot()
    })
    it('Should render the given month title with year if it is not the current year', () => {
      const futureYearProps = { ...captionProps, date: new Date(2050, 0, 1) }
      const caption = renderer.create(datepicker.instance().renderCaption(futureYearProps))
      expect(caption).toMatchSnapshot()
    })
  })

  describe('onChange', () => {
    it('Should return the date in the format `YYYY-MM-DD`', () => {
      const onChange = jest.fn()
      const date = new Date(2020, 0, 1)
      const wrapper = shallow(<DatePicker name="datepicker" onChange={onChange} />)
      wrapper.instance().onDayClick(date, {})
      expect(onChange).toHaveBeenCalledWith({ name: 'datepicker', value: '2020-01-01' })
    })
  })

  describe('react-day-picker', () => {
    it('Should forward props to the DayPicker component', () => {
      const now = new Date()
      const wrapper = mount(<DatePicker name="datepicker" initialDate={now} />)
      const instance = wrapper.instance()

      expect(wrapper.find(DayPicker).prop('numberOfMonths')).toEqual(2)
      expect(wrapper.find(DayPicker).prop('selectedDays')).toEqual(now)
      expect(wrapper.find(DayPicker).prop('onDayClick')).toEqual(instance.onDayClick)
      expect(wrapper.find(DayPicker).prop('navbarElement')).toEqual(instance.renderNavbar)
      expect(wrapper.find(DayPicker).prop('captionElement')).toEqual(instance.renderCaption)
      expect(wrapper.find(DayPicker).prop('renderDay')).toEqual(instance.renderDay)
      expect(wrapper.find(DayPicker).prop('firstDayOfWeek')).toEqual(0)
    })

    it('Should forward same value for selectedDays and initialMonth props to the DayPicker component', () => {
      const initialDate = new Date()
      initialDate.setMonth(initialDate.getMonth() + 6)
      const wrapper = mount(
        <DatePicker
          name="datepicker"
          initialDate={initialDate}
          initialMonth={initialDate}
          numberOfMonths={12}
        />,
      )

      expect(wrapper.find(DayPicker).prop('numberOfMonths')).toEqual(12)
      expect(wrapper.find(DayPicker).prop('selectedDays')).toEqual(initialDate)
      expect(wrapper.find(DayPicker).prop('initialMonth')).toEqual(initialDate)
    })

    it('Should forward only selectedDays and keep default initialMonth props to the DayPicker component', () => {
      const offsetTop = 1000
      const stickyPositionTop = 50
      window.scrollTo = jest.fn()
      Object.defineProperty(HTMLElement.prototype, 'offsetTop', {
        configurable: true,
        value: offsetTop,
      })

      const initialDate = new Date()
      initialDate.setMonth(initialDate.getMonth() + 6)
      const wrapper = mount(
        <DatePicker
          name="datepicker"
          initialDate={initialDate}
          numberOfMonths={12}
          stickyPositionTop={stickyPositionTop}
        />,
      )

      expect(wrapper.find(DayPicker).prop('numberOfMonths')).toEqual(12)
      expect(wrapper.find(DayPicker).prop('selectedDays')).toEqual(initialDate)
      expect(wrapper.find(DayPicker).prop('initialMonth')).not.toEqual(initialDate)
      expect(window.scrollTo).toHaveBeenCalledWith(0, offsetTop + stickyPositionTop)
    })

    afterEach(() => {
      // Reset offsetTop
      Object.defineProperty(HTMLElement.prototype, 'offsetTop', {
        configurable: true,
        value: 0,
      })

      jest.clearAllMocks()
    })
  })
})
