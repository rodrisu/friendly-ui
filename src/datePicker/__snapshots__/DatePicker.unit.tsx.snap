// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker renderCaption Should render the given month title 1`] = `
<div
  className="kirk-datepicker-month-caption DayPicker-Caption"
>
  January 
</div>
`;

exports[`DatePicker renderCaption Should render the given month title with year if it is not the current year 1`] = `
<div
  className="kirk-datepicker-month-caption DayPicker-Caption"
>
  January 2050
</div>
`;

exports[`DatePicker renderNavbar Should render the previous/next buttons in horizontal mode 1`] = `
.c1 {
  fill: #00AFF5;
}

.c1.kirk-icon-wrapper {
  display: inline-block;
  position: relative;
}

.c1.kirk-icon-shadowed {
  border-radius: 50%;
  box-shadow: 0 0 2px rgba(0,0,0,.3);
}

.c1 .kirk-icon-badge {
  position: absolute;
  bottom: 12px;
  left: 12px;
}

.c0 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 0 24px;
  border: 1px solid transparent;
  border-radius: 48px;
  font-size: 16px;
  line-height: 1;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  font-family: inherit;
  height: 48px;
  min-width: 48px;
  max-width: 100%;
  overflow: hidden;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: max-width 500ms ease-in, background-color 200ms ease-in;
  transition: max-width 500ms ease-in, background-color 200ms ease-in;
  white-space: nowrap;
}

.c0 svg,
.c0 img {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c0 svg {
  margin-right: 8px;
}

.c0.kirk-button-bubble svg {
  margin: 0;
}

.c0.kirk-itemChoice {
  border-radius: 0;
  overflow: visible;
  width: 100%;
}

.c0:hover,
.c0:active {
  outline: 0;
}

.c0.kirk-button-unstyled {
  background-color: transparent;
  border-radius: inherit;
  color: #00AFF5;
  padding: inherit;
  display: inherit;
  line-height: inherit;
  font-size: inherit;
  text-align: left;
  min-width: inherit;
  min-height: inherit;
  height: auto;
  white-space: normal;
}

.c0[disabled] {
  opacity: 0.5;
  cursor: default;
}

.c0.kirk-button-unstyled[disabled] {
  opacity: 1;
}

.c0.kirk-button-primary {
  background-color: #00AFF5;
  color: #FFF;
}

.c0.kirk-button-shadowed {
  border: 2px solid #FFF;
  box-shadow: 0 2px 4px rgba(0,0,0,.3);
  box-sizing: border-box;
}

.c0.kirk-button-primary:hover,
.c0.kirk-button-primary:focus,
.c0.kirk-button-primary:active {
  background-color: #008FC1;
}

.c0.kirk-button-secondary {
  background-color: #FFF;
  border-color: #DDD;
  color: #00AFF5;
}

.c0.kirk-button-secondary.kirk-button-shadowed {
  border: none;
}

.c0.kirk-button-secondary:hover,
.c0.kirk-button-secondary:focus,
.c0.kirk-button-secondary:active {
  background-color: #DDD;
}

.c0.kirk-button-tertiary {
  background-color: #FFF;
  border-color: transparent;
  color: #00AFF5;
}

.c0.kirk-button-tertiary:hover,
.c0.kirk-button-tertiary:focus,
.c0.kirk-button-tertiary:active {
  background-color: #DDD;
}

.c0.kirk-button-loading {
  background-color: transparent;
  max-width: initial;
  border: 0;
}

.c0.kirk-button-loading:hover,
.c0.kirk-button-loading:focus,
.c0.kirk-button-loading:active {
  background-color: transparent;
}

.c0.kirk-button-loading[disabled],
.c0.kirk-button-checked[disabled] {
  opacity: 1;
}

.c0 svg.kirk-button-loader {
  position: absolute;
  top: 0;
  left: 50%;
  margin-left: -24px;
}

.c0.kirk-button-checked {
  border: 0;
}

.c0.kirk-button-warning {
  background-color: #F53F5B;
  color: #FFF;
}

.c0.kirk-button-bubble {
  display: inline-block;
  text-align: center;
  padding: 0;
  line-height: 0;
  width: 48px;
  height: 48px;
}

.c0.kirk-button-bubble.kirk-button-shadowed {
  width: calc(48px - 2px * 2);
  height: calc(48px - 2px * 2);
}

.c0.kirk-button-facebook {
  background-color: #1877F2;
  color: #FFF;
}

.c0.kirk-button-vkontakte {
  background-color: #2787F5;
  color: #FFF;
}

@media (hover:none),(hover:on-demand) {
  .c0.kirk-button-secondary:hover {
    background-color: #FFF;
  }
}

<div
  className=""
>
  <button
    className="kirk-button kirk-button-tertiary kirk-button-bubble c0 kirk-datepicker-previous-month"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="kirk-icon c1"
      height={24}
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      
      <g
        fill="none"
        stroke="#00AFF5"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeMiterlimit="10"
        strokeWidth="2"
      >
        <path
          d="M9 19l-7-7 7-7"
        />
        <path
          d="M22 12H2"
        />
      </g>
    </svg>
  </button>
  <button
    className="kirk-button kirk-button-tertiary kirk-button-bubble c0 kirk-datepicker-next-month"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="kirk-icon c1"
      height={24}
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      
      <g
        fill="none"
        stroke="#00AFF5"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeMiterlimit="10"
        strokeWidth="2"
        transform="rotate(180 12 12)"
      >
        <path
          d="M9 19l-7-7 7-7"
        />
        <path
          d="M22 12H2"
        />
      </g>
    </svg>
  </button>
</div>
`;

exports[`DatePicker renderNavbar Should render the weekdays in vertical mode 1`] = `
<div
  aria-hidden="true"
  className="kirk-datepicker-sticky-weekdays"
  style={
    Object {
      "top": "0px",
    }
  }
>
  <div
    className="kirk-datepicker-sticky-weekdaysrow"
  >
    <div
      className="kirk-datepicker-sticky-weekday"
    >
      Su
    </div>
    <div
      className="kirk-datepicker-sticky-weekday"
    >
      Mo
    </div>
    <div
      className="kirk-datepicker-sticky-weekday"
    >
      Tu
    </div>
    <div
      className="kirk-datepicker-sticky-weekday"
    >
      We
    </div>
    <div
      className="kirk-datepicker-sticky-weekday"
    >
      Th
    </div>
    <div
      className="kirk-datepicker-sticky-weekday"
    >
      Fr
    </div>
    <div
      className="kirk-datepicker-sticky-weekday"
    >
      Sa
    </div>
  </div>
</div>
`;
